### PR TITLE
Feature/909 proxy delegatecall

### LINKED ISSUE #909: Developers SBAT check contract code existence in Proxy contract
### Expected Behavior
Use `EXTCODESIZE` to check contract existence before `delegatecall` in Proxy.sol

### Current Behavior

No check

### PR DESCRIPTION
**Pull request by r00ted199**
### Description

The intention of this PR is to validate if an address is a contract or EOA before calling the `delegatecall`

### Tested
 
Test added in `proxy.ts` calling `_setAndInitializeImplementation` with a EOA address.

### Other changes

The `isContract` function shouldn't be called from a constructor. Calling this function from a constructor will return 0 independently if the address given as parameter is a contract or EOA

### Related issues

- Fixes #909 , #910, #911 

### Backwards compatibility

_Brief explanation of why these changes are/are not backwards compatible._

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -206,6 +206,7 @@ jobs:
             ./scripts/ci_check_if_test_should_run_v2.sh ${FILES_TO_CHECK}
       - run:
           name: test
+          no_output_timeout: 20m
```
Comment (m-chrzan) [inline .circleci/config.yml]: Why is this change necessary?

File: packages/protocol/contracts/common/Proxy.sol [hunk 2]
```diff
@@ -88,6 +88,7 @@ contract Proxy {
     payable
     onlyOwner
   {
+    require(isContract(implementation), "Invalid contract address");
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/Proxy.sol]: Wouldn't this need to be checked on every proxied call, in case the implementation contract self destructs?

File: packages/protocol/contracts/common/Proxy.sol [hunk 3]
```diff
@@ -138,4 +139,20 @@ contract Proxy {
...
+
+
+  /**
+   * @dev isContract detect whether the address is 
+   *      a contract address or externally owned account (EOA)
+   * WARNING: Calling this function from a constructor will return 0
+  *           independently if the address given as parameter is a contract or EOA
+   * @return true if it is a contract address
+   */
+  function isContract(address addr) public view returns (bool) {
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/Proxy.sol]: We've been making all functions on the proxy prefixed with `_` to avoid potential clashes with implementation contract function names.

File: packages/protocol/contracts/common/Proxy.sol [hunk 4]
```diff
@@ -138,4 +139,20 @@ contract Proxy {
...
+    uint size;
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/Proxy.sol]: Nit: by convention, we use `uint256` throughout the code base.

Comment (aaitor) [reply packages/protocol/contracts/common/Proxy.sol]: :+1:

File: packages/protocol/contracts/common/Proxy.sol (hunk 3 above)
Comment (r00ted199) [reply packages/protocol/contracts/common/Proxy.sol]: :+1:

File: packages/protocol/contracts/common/Proxy.sol (hunk 4 above)
Comment (r00ted199) [reply packages/protocol/contracts/common/Proxy.sol]: :+1:

File: .circleci/config.yml (hunk 1 above)
Comment (r00ted199) [reply .circleci/config.yml]: was increased because timeouts were found in CircleCI, I can remove it

File: packages/protocol/contracts/common/Proxy.sol (hunk 2 above)
Comment (r00ted199) [reply packages/protocol/contracts/common/Proxy.sol]: added to fallback function

File: packages/protocol/contracts/common/Proxy.sol [hunk 5]
```diff
@@ -32,8 +32,16 @@ contract Proxy {
...
+
+    assembly {
+      implementationAddress := sload(implementationPosition)
+    }
+
+    require(_isContract(implementationAddress), "Invalid contract address");
+
     assembly {
-      let implementationAddress := sload(implementationPosition)
+      //let implementationAddress := sload(implementationPosition)
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/Proxy.sol]: Remove commented out code.

File: packages/protocol/contracts/common/Proxy.sol [hunk 6]
```diff
@@ -138,4 +147,20 @@ contract Proxy {
     }
     emit OwnerSet(newOwner);
   }
+
+
+  /**
+   * @dev _isContract detect whether the address is 
+   *      a contract address or externally owned account (EOA)
+   * WARNING: Calling this function from a constructor will return 0
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/Proxy.sol]: Nit: function returns `bool`, so I guess to be precise this should say `false`, not `0`?

File: packages/protocol/contracts/common/Proxy.sol [hunk 7]
```diff
@@ -138,4 +147,20 @@ contract Proxy {
...
+  *           independently if the address given as parameter is a contract or EOA
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/Proxy.sol]: Nit: `*` is not aligned :P

File: packages/protocol/contracts/common/Proxy.sol [hunk 8]
```diff
@@ -138,4 +147,20 @@ contract Proxy {
...
+   * @return true if it is a contract address
+   */
+  function _isContract(address addr) public view returns (bool) {
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/Proxy.sol]: This can be `internal` rather than `public`, right?

Comment (r00ted199) [reply packages/protocol/contracts/common/Proxy.sol]: :+1:

File: packages/protocol/contracts/common/Proxy.sol (hunk 7 above)
Comment (r00ted199) [reply packages/protocol/contracts/common/Proxy.sol]: :+1:

File: packages/protocol/contracts/common/Proxy.sol (hunk 6 above)
Comment (r00ted199) [reply packages/protocol/contracts/common/Proxy.sol]: :+1:

File: packages/protocol/contracts/common/Proxy.sol (hunk 5 above)
Comment (r00ted199) [reply packages/protocol/contracts/common/Proxy.sol]: :+1:

**Comment by m-chrzan** [comment]
Transfer tests are failing. They are coupled to gas costs of various smart contract calls. Changing Proxy's fallback function will have increased gas costs for pretty much every call. This will unfortunately require changes to the e2e transfer test (updating the expected gas costs) and to `params/protocol_params.go` in celo-blockchain, to pass enough gas to the EVM when making internal EVM calls.

You can run the tests locally by going to the celotool package and running

    ./ci_test_transfers.sh local <path to local celo-blockchain repo>

Let me know if you have any questions, happy to chat about this.

File: packages/protocol/contracts/common/Proxy.sol [hunk 9]
```diff
@@ -88,6 +95,7 @@ contract Proxy {
     payable
     onlyOwner
   {
+    require(_isContract(implementation), "Invalid contract address");
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/Proxy.sol]: Actually, let's move this check to `_setImplementation` to catch this both at initialization time, and any other time the implementation changes.

File: packages/protocol/contracts/common/Proxy.sol [hunk 10]
```diff
@@ -32,8 +33,16 @@ contract Proxy {
   function () external payable {
     bytes32 implementationPosition = IMPLEMENTATION_POSITION;
 
+    address implementationAddress;
+
+    assembly {
+      implementationAddress := sload(implementationPosition)
+    }
+
+    require(
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/Proxy.sol]: Style: if line needs to be split (over 100 characters), each argument on new line, put `);` on new line.

File: packages/protocol/contracts/common/Proxy.sol [hunk 11]
```diff
@@ -114,6 +123,9 @@ contract Proxy {
   function _setImplementation(address implementation) public onlyOwner {
     bytes32 implementationPosition = IMPLEMENTATION_POSITION;
 
+    require(
+      AddressesHelper.isContract(implementation), "Invalid contract address");
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/Proxy.sol]: Ditto.

File: packages/protocol/contracts/common/Proxy.sol [hunk 12]
```diff
@@ -138,4 +150,5 @@ contract Proxy {
     }
     emit OwnerSet(newOwner);
   }
+
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/Proxy.sol]: Remove empty line at end of block.

File: packages/protocol/contracts/common/libraries/AddressesHelper.sol [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+   *          independently if the address given as parameter is a contract or EOA
+   * @return true if it is a contract address
+   */
+  function isContract(address addr) internal view returns (bool) {
+    uint256 size;
+    /* solium-disable-next-line security/no-inline-assembly */
+    assembly { size := extcodesize(addr) }
+    return size > 0;
+  }
+
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/libraries/AddressesHelper.sol]: Ditto.

File: packages/protocol/contracts/common/MultiSig.sol [hunk 14]
```diff
@@ -261,6 +262,10 @@ contract MultiSig is Initializable {
     returns (bool)
   {
     bool result;
+
+    require(
+      AddressesHelper.isContract(destination), "Invalid contract address");
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/MultiSig.sol]: Another `);` here.

File: packages/protocol/contracts/governance/Governance.sol [hunk 15]
```diff
@@ -1062,6 +1063,10 @@ contract Governance is IGovernance, Ownable, Initializable, UsingLockedGold, Ree
     returns (bool)
   {
     bool result;
+
+    require(
+      AddressesHelper.isContract(destination), "Invalid contract address");
```
Comment (m-chrzan) [inline packages/protocol/contracts/governance/Governance.sol]: Another `);`.

File: packages/protocol/test/governance/governance.ts [hunk 16]
```diff
@@ -1448,6 +1448,29 @@ contract('Governance', (accounts: string[]) => {
           await assertRevert(governance.execute(proposalId, index))
         })
       })
+
+      describe('when the proposal cannot execute successfully because is not a contract address', () => {
```
Comment (m-chrzan) [inline packages/protocol/test/governance/governance.ts]: Typo: "because *it* is not..."

File: .circleci/config.yml [hunk 17]
```diff
@@ -206,6 +206,7 @@ jobs:
             ./scripts/ci_check_if_test_should_run_v2.sh ${FILES_TO_CHECK}
       - run:
           name: test
+          # no_output_timeout: 20m
```
Comment (m-chrzan) [inline .circleci/config.yml]: If this was necessary, leave it in. Otherwise remove commented out code.

**Review by m-chrzan — APPROVED** [review]
Looks good, just some style nits. Removed `automerge` so you can address them first.

**Comment by r00ted199** [comment]
@asaj, @m-chrzan  , this PR is pending of some broken tests in `master`. Let me know if is good to merge or you prefer till those changes are fixed there first.

File: packages/protocol/contracts/governance/Governance.sol [hunk 18]
```diff
@@ -1052,4 +1053,50 @@ contract Governance is IGovernance, Ownable, Initializable, UsingLockedGold, Ree
     }
     return threshold;
   }
+
+  // call has been separated into its own function in order to take advantage
```
Comment (asaj) [inline packages/protocol/contracts/governance/Governance.sol]: Merge conflict? I believe this was recently moved into `Proposals.sol`

File: packages/protocol/contracts/governance/Governance.sol [hunk 19]
```diff
@@ -1052,4 +1053,50 @@ contract Governance is IGovernance, Ownable, Initializable, UsingLockedGold, Ree
...
+    uint dataLength,
+    bytes memory data
+  )
+    private
+    nonReentrant
+    returns (bool)
+  {
+    bool result;
+
+    require(AddressesHelper.isContract(destination), "Invalid contract address");
```
Comment (asaj) [inline packages/protocol/contracts/governance/Governance.sol]: Do we want to always require this? Or only when dataLength is > 0?

I would imagine in some cases we will have governance proposals to send gold or dollars to an EOA.

File: packages/protocol/contracts/common/MultiSig.sol [hunk 20]
```diff
@@ -261,6 +262,9 @@ contract MultiSig is Initializable {
     returns (bool)
   {
     bool result;
+
+    require(AddressesHelper.isContract(destination), "Invalid contract address");
```
Comment (asaj) [inline packages/protocol/contracts/common/MultiSig.sol]: Ditto on the `dataLength` question

**Review by asaj — CHANGES_REQUESTED** [review]

**Comment by asaj** [comment]
> @asaj, @m-chrzan , this PR is pending of some broken tests in master. Let me know if is good to merge or you prefer till those changes are fixed there first.

I think it's worth waiting to fix these, I think the fix should be quick and simple

File: packages/protocol/contracts/common/MultiSig.sol (hunk 20 above)
Comment (r00ted199) [reply packages/protocol/contracts/common/MultiSig.sol]: :+1:

File: packages/protocol/contracts/governance/Governance.sol (hunk 19 above)
Comment (r00ted199) [reply packages/protocol/contracts/governance/Governance.sol]: :+1:

File: packages/protocol/contracts/governance/Governance.sol (hunk 18 above)
Comment (r00ted199) [reply packages/protocol/contracts/governance/Governance.sol]: yep, fixed

**Review by asaj — APPROVED** [review]
